### ISSUE TITLE
Inconsistent matrix-vector product for jit and non-jit version on cpu. 

### ISSUE THREAD

**Issue by abhiagwl** [google/jax#5381]
For some special matrices, the matrix-vector product is inconsistent when calculated on the CPU. The same computation is consistent on GPU.
```python
    import jax.numpy as np
    import numpy as onp

    M = np.array([
            [9.26, 0, 0, 0],
            [1.02e+04, 11.6, 0, 0],
            [-2.92e+05, -331, 0.73, 0],
            [2.35e+07, 2.66e+04, -58.4, 5]])
    v = np.array([ -0.2, 172.9,    -34,      -3 ])

    m_1 = lambda M, v: np.dot(M, v)
    j_m_1 = jax.jit(m_1)
    m_2 = lambda : np.dot(M, v)
    j_m_2 = jax.jit(m_2)
    m_3 = onp.dot


    print(f"Method 1, jitted vs. non jitted: {j_m_1(M,v) - m_1(M, v)}")
    print(f"Method 1, jitted vs. Method 2, jitted: {j_m_1(M, v) - j_m_2()}")
    print(f"Method 1, jitted vs. Method 3: {j_m_1(M, v) - m_3(M,v)}")
    print("\n########################################################\n")

    print(f"Method 2, jitted vs. non jitted: {j_m_2() - m_2()}")
    print(f"Method 2, jitted vs. Method 3: {j_m_2() - m_3(M,v)}")
```

On CPU (32 bit):
```
Method 1, jitted vs. non jitted: [0. 0. 0. 0.]
Method 1, jitted vs. Method 2, jitted: [ 0.0000000e+00 -1.9073486e-05  4.8828125e-04 -1.6406250e-01]
Method 1, jitted vs. Method 3: [ 0.0000000e+00 -1.9073486e-05  4.8828125e-04 -1.6406250e-01]

########################################################

Method 2, jitted vs. non jitted: [ 0.0000000e+00  1.9073486e-05 -4.8828125e-04  1.6406250e-01]
Method 2, jitted vs. Method 3: [0. 0. 0. 0.]
```

On CPU (64 bit):
```
Method 1, jitted vs. non jitted: [0. 0. 0. 0.]
Method 1, jitted vs. Method 2, jitted: [0. 0. 0. 0.]
Method 1, jitted vs. Method 3: [ 0.00000000e+00  0.00000000e+00 -2.27373675e-13  3.78349796e-10]

########################################################

Method 2, jitted vs. non jitted: [0. 0. 0. 0.]
Method 2, jitted vs. Method 3: [ 0.00000000e+00  0.00000000e+00 -2.27373675e-13  3.78349796e-10]
```

On GPU (32 bit):
```
Method 1, jitted vs. non jitted: [0. 0. 0. 0.]
Method 1, jitted vs. Method 2, jitted: [0. 0. 0. 0.]
Method 1, jitted vs. Method 3: [0. 0. 0. 0.]

########################################################

Method 2, jitted vs. non jitted: [0. 0. 0. 0.]
Method 2, jitted vs. Method 3: [0. 0. 0. 0.]
```
The original numpy evaluation is the same for both CPU and GPU(32 bit.) I am also aware that matrix M is ill-conditioned; however, it seems strange that it should interfere with matrix multiplication. I am using jax 0.2.7 and jaxlib 0.1.56.

**Comment by zhangqiaorjc** [comment]
This may be a dtype issue where JAX prefers float32 and Numpy float64. Jake may know the dtypes rules more.

**Comment by jakevdp** [comment]
I think this is simply floating point error accumulating due to different order of operations. With some trial and error I was able to replicate the same diffs using numpy alone, by shuffling the order of the columns:
```python
import numpy as np
M = np.array([
        [9.26, 0, 0, 0],
        [1.02e+04, 11.6, 0, 0],
        [-2.92e+05, -331, 0.73, 0],
        [2.35e+07, 2.66e+04, -58.4, 5]], dtype='float32')
v = np.array([ -0.2, 172.9,    -34,      -3 ], dtype='float32')
i = np.array([0, 2, 1, 3])

np.dot(M,v) - np.dot(M[:, i], v[i])
# array([ 0.0000000e+00,  1.9073486e-05, -4.8828125e-04,  1.6406250e-01],
#       dtype=float32)
```
When JAX's JIT optimizes a computation, it often involves re-ordering floating point operations, which can change the details of the result withing standard floating point precision; I think that's what's happening here.
